### PR TITLE
[PyTorch][Fused Attn] Add support for cuDNN to return Softmax `Stats` always and `Max` when `return_max_logit=True`

### PR DESCRIPTION
**Pull request by sudhakarsingh27**
# Description
cuDNN recently made returning any subset of {Stats, SumExp, Max} possible. This PR adapts TE to always get `Stats` from cuDNN and `Max` tensor if `return_max_logit=True`. (Note that `Stats` = log(SumExp)+Max)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:
-  `fused_attn_f16_arbitrary_seqlen.cu`
    - Removed references to `SumExp` tensor as it's not needed since cuDNN returns `Stats` by default.
    - set `generate_stats=True` which forces cuDNN to always return `Stats` tensor (needed in the backward pass)
- `transformer_engine/pytorch/cpp_extensions/fused_attn.py`
    - Remove code that manually did `Stats = log(SumExp) + Max` since cuDNN returns `Stats` directly and TE doesn't need `SumExp` from cuDNN 
- Corresponding documentation 

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/cpp_extensions/fused_attn.py [hunk 1]
```diff
@@ -339,14 +339,14 @@ def fused_attn_fwd(
 
     if return_max_logit:
         qkv_format = qkv_layout.replace("3", "").replace("2", "").split("_")[0]
-        # thd:  output_tensors: out [tq, h, d],    Max [tq, h, 1],    Sum_Exp [tq, h, 1]
-        # bshd: output_tensors: out [b, sq, h, d], Max [b, h, sq, 1], Sum_Exp [b, h, sq, 1]
-        # sbhd: output_tensors: out [sq, b, h, d], Max [b, h, sq, 1], Sum_Exp [b, h, sq, 1]
-        stats = output_tensors[1] + torch.log(output_tensors[2])
+        # thd:  output_tensors: out [tq, h, d],    Stats [tq, h, 1],    Max [tq, h, 1]
+        # bshd: output_tensors: out [b, sq, h, d], Stats [b, h, sq, 1], Max [b, h, sq, 1]
+        # sbhd: output_tensors: out [sq, b, h, d], Stats [b, h, sq, 1], Max [b, h, sq, 1] (there's no typo here)
```
Comment (cyanguwa) [inline transformer_engine/pytorch/cpp_extensions/fused_attn.py]: Do we need the "there's no typo here" :)

File: transformer_engine/pytorch/csrc/extensions/attention.cpp [hunk 2]
```diff
@@ -259,16 +259,16 @@ std::vector<py::object> fused_attn_fwd(
   // f16_max512   : S [b, h, sq, skv]
   // f16_arbitrary:
   // return_max_logit=false: S [b, h, sq, 1], rng_state [2], (optional) Bias [1, h, sq, skv], (optional) SoftmaxOffset [1, h, 1, 1]
-  // return_max_logit=true: Max [b, h, sq, 1], Sum_Exp [b, h, sq, 1], rng_state [2], (optional) Bias [1, h, sq, skv], (optional) SoftmaxOffset [1, h, 1, 1]
+  // return_max_logit=true: S [b, h, sq, 1], Max [b, h, sq, 1], rng_state [2], (optional) Bias [1, h, sq, skv], (optional) SoftmaxOffset [1, h, 1, 1]
   // fp8          : M [b, h, sq, 1], ZInv [b, h, sq, 1], rng_state [2]
   size_t i = 0;
   at::Tensor output_tensor;
-  // intermediate softmax tensor, S or M
+  // intermediate softmax tensor, S (or `Stats`)
```
Comment (cyanguwa) [inline transformer_engine/pytorch/csrc/extensions/attention.cpp]: The "M" is for the FP8 case. I understand your change but I think we probably don't need it.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 3]
```diff
@@ -1116,6 +1107,16 @@ void fused_attn_arbitrary_seqlen_fwd(
   size_t i = 0;
   if (Aux_CTX_Tensors->size == 0) {
     const auto cudnn_runtime_version = cudnnGetVersion();
+
```
Comment (cyanguwa) [inline transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:1131]: You might need to make these changes in the "Aux_CTX_Tensors->size == 0" sections in _fwd/bwd_qkvpacked/kvpacked APIs as well. Please check. Thanks!

File: transformer_engine/pytorch/cpp_extensions/fused_attn.py (hunk 1 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/cpp_extensions/fused_attn.py]: I deliberately added it because I didn't believe it and checked the shapes myself :P

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 4]
```diff
@@ -103,7 +103,7 @@ void fused_attn_arbitrary_seqlen_fwd_impl(
   }
 
   const DType ragged_offset_type = cudnn_runtime_version >= 90500 ? DType::kInt64 : DType::kInt32;
-  bool generate_stats = !return_max_logit;
+  bool generate_stats = true;  // Always return stats
```
[the remark this answers is not in the document]
Comment (KshitijLakhani) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:115]: Good catch by greptile to change commetns in `fused_attn.h`

Also, @sudhakarsingh27 do we need the ` bool generate_max_sum_exp` still in `utils.h` ?

File: transformer_engine/pytorch/cpp_extensions/fused_attn.py [hunk 5]
```diff
@@ -339,14 +339,14 @@ def fused_attn_fwd(
...
+
+        aux_ctx_tensors = [output_tensors[1]]  # "Stats"
         amax_dims = (0, 2) if qkv_format == "thd" else (0, 2, 3)
         # Max -> max_logit [h]
-        max_logit = torch.amax(output_tensors[1], dim=amax_dims).to(dtype=output_tensors[0].dtype)
-        aux_ctx_tensors = [stats]
+        max_logit = torch.amax(output_tensors[2], dim=amax_dims).to(dtype=output_tensors[0].dtype)
```
Comment (KshitijLakhani) [inline transformer_engine/pytorch/cpp_extensions/fused_attn.py]: Maybe I understood this incorrectly, but isn't TE now also suppose to receive max from cuDNN directly (like stats, but with stats it is always true and with max it cn be toggled) rather than calling amax() in TE ?


(Sudhakar: Why am I able to update your comment? )

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/cpp_extensions/fused_attn.py]: cuDNN returns Max ([b, h, sq, 1]) so it's an additional softmax statistic (apparently, the subset (`Stats`, `Max`) is enough for cuDNN bwd rather than the full set `(Stats, SumExp, Max)`). 

Further, for muon, we need do amax on it to get a dimension  [h] tensor. `return_max_logit` in TE controls whether to fetch `Max` from cuDNN. 

Perf wise, it'd be nice for cuDNN to do additional reduction to return the [h] shaped tensor for muon as well but that's not  the scope of this PR. 

(Kshitij: looks like I can as well)

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu (hunk 4 above)
[the remark this answers is not in the document]
Comment (sudhakarsingh27) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:115]: I wasn't even aware of this, thanks for pointing it out. I think this should just be renamed to `return_max_logit`.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu (hunk 3 above)
Comment (sudhakarsingh27) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:1131]: Looks like I don't need to because `nvte_fused...qvpacked` are in `fused_attn.cpp` which calls `fused_attn_f16_arbitrary...` just like regular nvte_fused_fwd/bwd

File: transformer_engine/pytorch/cpp_extensions/fused_attn.py [hunk 6]
```diff
@@ -353,24 +353,24 @@ def fused_attn_fwd(
...
             seqlens_q = (cu_seqlens_q[1:] - cu_seqlens_q[:-1]).to(device=max_tensor.device)
             sq_idx = torch.arange(max_tensor.shape[2], device=max_tensor.device).view(1, 1, -1, 1)
             valid = sq_idx < seqlens_q.view(-1, 1, 1, 1)
             max_tensor = max_tensor.masked_fill(~valid, float("-inf"))
-        amax_dims = (0, 2) if max_tensor.ndim == 3 else (0, 2, 3)
+
         # Max -> max_logit [h]
         max_logit = torch.amax(max_tensor, dim=amax_dims).to(dtype=output_tensors[0].dtype)
-        aux_ctx_tensors = [stats]
-        aux_ctx_tensors.extend(output_tensors[3:])
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/cpp_extensions/fused_attn.py:373]: That seems like an oversight

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/cpp_extensions/fused_attn.py:373]: Fixed with `aux_ctx_tensors = [output_tensors[1]] + list(output_tensors[3:])  # Stats + rng_state + optional tensors`

**Review by cyanguwa — APPROVED** [review]
